### PR TITLE
Feature/different approach for store item

### PR DESCRIPTION
**Pull request by dieDeMiguel**
This branch implements logic for all the routes in the handler. Also adds new utils in order to santize and validate code.

### CODE REVIEW

File: controllers/items.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+          } else {
+              callback(null, response);
+          }
+       });
+    } else {
+        callback("The item must have a numeric ID and a name larger than 3 words", null)
+    }
+}
+
+const readFileFromMemory = (fileName, callback) => {
```
Comment (charlitos85) [inline controllers/items.js:21]: Const listItems

File: controllers/items.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            if(error) {
+                callback(error, null)
+            } else {
+                callback(null, response);
+            }
+        })
+    }
+}
+
+const getByID = (fileName, id, callback) => {
```
Comment (charlitos85) [inline controllers/items.js:33]: Const getItem

File: routes/index.js [hunk 3]
```diff
@@ -1,31 +1,41 @@
...
   res.render('index', { title: 'Express' });
 });
 
 
-// GET objects array complete
-router.get('/items', function(req, res, next) {
-  res.send('Here you will find the array parsed');
+// GET items 
+router.get('/items', function(req, res) {
+    const fileName = req.body.fileName;
```
Comment (charlitos85) [inline routes/index.js:13]: Move this to a constant

File: controllers/items.js (hunk 1 above)
Comment (dieDeMiguel) [reply controllers/items.js:21]: done and also updated every time this function was called

File: controllers/items.js (hunk 2 above)
Comment (dieDeMiguel) [reply controllers/items.js:33]: done and also updated every time this function was called

File: routes/index.js (hunk 3 above)
Comment (dieDeMiguel) [reply routes/index.js]: done

File: utils/storeitem.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const fs = require('fs');
+const fileName = 'items.csv';
```
Comment (charlitos85) [inline utils/storeitem.js:2]: This should not be here

Comment (dieDeMiguel) [reply utils/storeitem.js:2]: done

File: controllers/items.js [hunk 5]
```diff
@@ -67,10 +68,22 @@ const updateItem = (itemToUpdate, fileName, callback) => {
         callback("The item must have a numeric ID and a name larger than 3 words", null)
     }
 }
+
+const removeFile = (callback) => {
```
Comment (charlitos85) [inline controllers/items.js:72]: this function should remove only a item

File: controllers/items.js [hunk 6]
```diff
@@ -18,20 +20,20 @@ const  createItem = (item, callback) =>{
...
-        readFile(fileName, (error, response) => {
-            if(error) {
-                callback(error, null)
-            } else {
-                callback(null, response);
-            }
-        })
-    }
+const listItems = (callback) => {
+    const fileName = 'items.csv';
```
Comment (charlitos85) [inline controllers/items.js:24]: move this outside the function

File: controllers/items.js [hunk 7]
```diff
@@ -46,14 +48,13 @@ const getItem = (fileName, id, callback) => {
 
 
 //Update Item route.
-const updateItem = (itemToUpdate, fileName, callback) => {
-    if(validateDataName(itemToUpdate.name, itemToUpdate.id)) {
-        fetchOneByID(fileName, itemToUpdate.id, (error, response) => {
+const updateItem = (itemToUpdate, id, callback) => {
```
Comment (charlitos85) [inline controllers/items.js:51]: const updateItem = (id, item, callback) => {

File: utils/validation.js [hunk 8]
```diff
@@ -12,17 +12,17 @@ function validateDataName(name, id) {
   } 
 }
 
-function validateDataFileName(fileName, id=1) {
-  fileName = fileName.trim();
+function validateID(id=1) {
   id = parseInt(id);
-  if(!Number.isInteger(id) || fileName.length <6) {
+  if(!Number.isInteger(id)) {
```
Comment (charlitos85) [inline utils/validation.js:17]: return Number.isInteger(id);

File: controllers/items.js [hunk 9]
```diff
@@ -67,10 +68,22 @@ const updateItem = (itemToUpdate, fileName, callback) => {
         callback("The item must have a numeric ID and a name larger than 3 words", null)
     }
 }
+
+const removeFile = (callback) => {
+    const fileName = "items.csv";
+    deleteFile(fileName, (error, response) => {
+        if(error) {
+            callback(error, null)
```
Comment (charlitos85) [inline controllers/items.js:76]: missing `;`. please review all the code and add the missing ones

File: controllers/items.js [hunk 10]
```diff
@@ -67,10 +68,22 @@ const updateItem = (itemToUpdate, fileName, callback) => {
...
+    })  
+}
  
 module.exports = {
     createItem: createItem,
     listItems: listItems,
     getItem: getItem,
-    updateItem: updateItem
+    updateItem: updateItem, 
+    removeFile: removeFile
```
Comment (charlitos85) [inline controllers/items.js:88]: deleteItem

File: routes/index.js [hunk 11]
```diff
@@ -1,7 +1,7 @@
 var express = require('express');
 var router = express.Router();
-const {createItem, listItems, getItem, updateItem } = require('../controllers/items');
-const fileName = 'items.csv';
+const {createItem, listItems, getItem, updateItem, removeFile } = require('../controllers/items');
```
Comment (charlitos85) [inline routes/index.js:3]: const ItemController = require('../controllers/items');

File: utils/deletefile.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+const fs = require('fs');
+
+function deleteFile(fileName, callback) {
+    fs.unlink(fileName, (err) => {
+        if (err) throw err;
```
Comment (charlitos85) [inline utils/deletefile.js:5]: you  should return the error in the callback. that is wrong

File: controllers/items.js (hunk 5 above)
Comment (dieDeMiguel) [reply controllers/items.js:72]: done, the function now looks into the file and deletes the item with the provided ID

Comment (dieDeMiguel) [reply controllers/items.js:72]: update: the function now  removes one item from the file

File: controllers/items.js (hunk 6 above)
Comment (dieDeMiguel) [reply controllers/items.js:24]: done

File: controllers/items.js (hunk 7 above)
Comment (dieDeMiguel) [reply controllers/items.js:51]: changed, now itemToUpdate is changes

File: controllers/items.js (hunk 10 above)
Comment (dieDeMiguel) [reply controllers/items.js:88]: done

File: routes/index.js (hunk 11 above)
Comment (dieDeMiguel) [reply routes/index.js:3]: done

File: utils/deletefile.js (hunk 12 above)
Comment (dieDeMiguel) [reply utils/deletefile.js:5]: done

File: controllers/items.js (hunk 9 above)
Comment (dieDeMiguel) [reply controllers/items.js]: done

File: utils/validation.js (hunk 8 above)
Comment (dieDeMiguel) [reply utils/validation.js]: done

File: utils/deletefile.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+const fs = require('fs');
+const {fetchOneByIDAndDelete}  = require('./getoneitem');
+
+function deleteItem(fileName, id, callback) {
+     fetchOneByIDAndDelete(fileName, id, (error, response) => {
+         if(error) {
```
Comment (charlitos85) [inline utils/deletefile.js:6]: when you don't do anything with the results of the callback, you can directly pass the callback

```if(error) {
            callback(error, null);
        } else {
            callback(null, response);
        }
```

to this

```
fetchOneByIDAndDelete(fileName, id, callback);
```

File: utils/storeitem.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+                            callback(error, null);
+                        } else {
+                            callback(null, 'The object with ID: #' + objectToStore.id + ' has been saved to memory');
+                            return;
+                        }
+                    });
+                }
+            })         
+        } else {
+            const firstItem = [ {"id":0, "name": "#"} ];
```
Comment (charlitos85) [inline utils/storeitem.js:25]: don't need this

File: utils/storeitem.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+                    callback(error, null)
+                } else {
+                    itemsArray = response;
+                    itemsArray.push(objectToStore);
+                    itemsArray = JSON.stringify(itemsArray);
+                    fs.writeFile(fileName, itemsArray, (error)=> {
+                        if(error) {
+                            callback(error, null);
+                        } else {
+                            callback(null, 'The object with ID: #' + objectToStore.id + ' has been saved to memory');
```
Comment (charlitos85) [inline utils/storeitem.js:18]: you need to return the obj

File: utils/storeitem.js [hunk 16]
```diff
@@ -0,0 +1,66 @@
+const fs = require('fs');
+const readFile = require('./readfile');
+
+function storeItem(objectToStore, fileName, callback) {
```
Comment (charlitos85) [inline utils/storeitem.js:4]: `objectToStore` is `item`

File: utils/storeitem.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    fs.access(fileName, (err) => {
+        if (!err) {
+            readFile(fileName, (error, response) => {
+                if(error) {
+                    callback(error, null)
+                } else {
+                    itemsArray = response;
+                    itemsArray.push(objectToStore);
+                    itemsArray = JSON.stringify(itemsArray);
```
Comment (charlitos85) [inline utils/storeitem.js:13]: its not responsabilty of the controller save as string, this should to be happen into the writeFile

File: utils/storeitem.js [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+                    callback(error);
+                    return;
+                } else {
+                    readFile(fileName, (error, response) => {
+                        if(error) {
+                            callback(error, null)
+                        } else {
+                            itemsArray = response;
+                            itemsArray.push(objectToStore);
+                            itemsArray = JSON.stringify(itemsArray);
```
Comment (charlitos85) [inline utils/storeitem.js:37]: same here

File: utils/storeitem.js [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    fs.writeFile(fileName, objectsToStore, (error) => {
+        if(error) {
+            callback('Something went wrong while storing the changes in the file');
+        } callback(null, 'The file was updated');
+    })
+}
+
+
+module.exports = {
+    storeItem: storeItem,
```
Comment (charlitos85) [inline utils/storeitem.js:64]: in ES6 you can directly add the name of the attribute

```
module.exports = {
   storeItem,
   ...
}
```

File: utils/validation.js [hunk 20]
```diff
@@ -0,0 +1,24 @@
+const fs = require('fs');
+  
+function validateDataName(name, id) {
+  id = parseInt(id);
+  name = name.trim();
+  if(id.length == 0 || name.length == 0) {
```
Comment (charlitos85) [inline utils/validation.js:6]: id.length == 0 ????

File: .gitignore [hunk 21]
```diff
@@ -1,4 +1,4 @@
 /node_modules
 package-lock.json
-items.txt
+items.csv
```
Comment (charlitos85) [inline .gitignore:3]: items.json

File: controllers/items.js [hunk 22]
```diff
@@ -0,0 +1,92 @@
+const fs = require('fs');
+const {storeItem} = require('../utils/storeitem');
+const readFile = require('../utils/readfile');
+const { validateDataName, validateID } = require('../utils/validation');
+const deleteItem = require('../utils/deletefile');
+const { fetchOneByID, fetchOneByIDAndDelete } = require('../utils/getoneitem');
+const fileName="items.csv"
```
Comment (charlitos85) [inline controllers/items.js:7]: items.json

File: routes/index.js [hunk 23]
```diff
@@ -1,31 +1,49 @@
...
+//Update an Item, req.body contains the changes in the item, the ID of the item is whithin the URL
+router.put('/items/:id', function(req, res) {
+  itemController.updateItem(req.body, req.params.id, (error, response) => {
+    error == null ? res.status(202).send(response) : res.status(400).send(error);
+  })
+})
+
+//Delete an Item in the file route
+router.delete('/items/:id', function(req, res) {
+  itemController.removeFile(req.params.id, (error, response) => {
```
[the remark this answers is not in the document]
Comment (dieDeMiguel) [reply routes/index.js]: done

File: utils/storeitem.js [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+            });
+        }
+    })
+} 
+
+function storeAfterDeleting(objectsToStore, fileName, callback) {
+    fs.writeFile(fileName, objectsToStore, (error) => {
+        if(error) {
+            callback('Something went wrong while storing the changes in the file');
+        } callback(null, 'The file was updated');
```
[the remark this answers is not in the document]
Comment (dieDeMiguel) [reply utils/storeitem.js]: storItem now checks wether or not the file exists. If exists adds the newItem and save. If it doesn't exist creates an empty array and push the newItem into it, then saves the array woth the object with writeFile

File: utils/deletefile.js (hunk 13 above)
Comment (dieDeMiguel) [reply utils/deletefile.js]: done

File: utils/storeitem.js (hunk 14 above)
Comment (dieDeMiguel) [reply utils/storeitem.js]: removed

File: utils/storeitem.js (hunk 16 above)
Comment (dieDeMiguel) [reply utils/storeitem.js]: changed

File: utils/storeitem.js (hunk 17 above)
Comment (dieDeMiguel) [reply utils/storeitem.js]: solved

File: utils/storeitem.js (hunk 18 above)
Comment (dieDeMiguel) [reply utils/storeitem.js]: solved

File: utils/storeitem.js (hunk 19 above)
Comment (dieDeMiguel) [reply utils/storeitem.js]: done

File: utils/validation.js (hunk 20 above)
Comment (dieDeMiguel) [reply utils/validation.js]: I added one = son now there is three of them checking equality

File: .gitignore (hunk 21 above)
Comment (dieDeMiguel) [reply .gitignore:3]: done

File: controllers/items.js (hunk 22 above)
Comment (dieDeMiguel) [reply controllers/items.js]: done

**Comment by dieDeMiguel** [comment]
CRUD completed

**Pull request merged**